File: nodenithy/run/public_key_service.js
const axios = require('axios');
const fs = require('fs');
const { Command } = require('commander');
const program = new Command();

const BASE_URL = "https://publickey.ethernity.cloud";

async function submitIpfsHash(hash, enclaveName, protocolVersion, network, templateVersion, dockerComposerHash) {
    const url = `${BASE_URL}/api/addHash`;
    const payload = {
        hash: hash,
        enclave_name: enclaveName,
        protocol_version: protocolVersion,
        network: network,
        template_version: templateVersion,
        docker_composer_hash: dockerComposerHash,
    };
    try {
        const response = await axios.post(url, payload);
        // console.log(`response: ${response}`);
        return response.data;
    } catch (error) {
        console.error("Error:", error.response ? error.response.data : error.message);
        process.exit(1);
    }
}

async function checkIpfsHashStatus(hash) {
    const url = `${BASE_URL}/api/checkHash/${hash}`;
    try {
        const response = await axios.get(url);
        return response.data;
    } catch (error) {
        console.error("Error:", error.response ? error.response.data : error.message);
        process.exit(1);
    }
}

program
    .requiredOption('--enclave_name <enclaveName>', 'Enclave name')
    .requiredOption('--protocol_version <protocolVersion>', 'Protocol version')
    .requiredOption('--network <network>', 'Network')
    .requiredOption('--template_version <templateVersion>', 'Template version');

program.parse(process.argv);

const options = program.opts();
const enclaveName = options.enclave_name;
const protocolVersion = options.protocol_version;
const network = options.network.toLowerCase().split("_")[1];
const templateVersion = options.template_version;

const hhash = fs.readFileSync('IPFS_HASH.ipfs', 'utf8').trim();
const dockerComposerHash = fs.readFileSync('IPFS_DOCKER_COMPOSE_HASH.ipfs', 'utf8').trim();

console.log("IPFS Hash:", hhash);
console.log("Enclave Name:", enclaveName);
console.log("Protocol Version:", protocolVersion);
console.log("Network:", network);
console.log("Template Version:", templateVersion);
console.log("Docker Composer Hash:", dockerComposerHash);

(async () => {
    // Submit IPFS Hash
    const submitResponse = await submitIpfsHash(hhash, enclaveName, protocolVersion, network, templateVersion, dockerComposerHash);
    console.log("Submit IPFS Hash Response:", submitResponse);

    // Check IPFS Hash Status
    while (true) {
        const checkResponse = await checkIpfsHashStatus(hhash);
        if ("publicKey" in checkResponse) {
            if (checkResponse.publicKey === 0) {
                console.log(`Public key not available yet. Queue position: ${checkResponse.queuePosition || 'Unknown'}`);
            } else if (checkResponse.publicKey === -1) {
                console.log("Hash is not derived from Eternity Cloud SDK.");
                process.exit(1);
            } else {
                console.log("Public Key:", checkResponse.publicKey);
                // Save public key to file
                fs.writeFileSync('PUBLIC_KEY.txt', checkResponse.publicKey);
                // ESR (RFC §5.2): the enclave also prints ESR_WALLET_ADDRESS on the
                // cert channel. The service currently returns only "publicKey"; when
                // it starts relaying the enclave's output (or adds an
                // "esrWalletAddress" field) it is picked up here and written to a
                // sibling file the caller reads -- no other changes needed. Until
                // then, ESR projects extract locally on an SGX host once to learn
                // the address (publish then persists it).
                if (checkResponse.esrWalletAddress) {
                    console.log("ESR Wallet Address:", checkResponse.esrWalletAddress);
                    fs.writeFileSync('ESR_WALLET_ADDRESS.txt', checkResponse.esrWalletAddress);
                }
                break;
            }
        } else {
            console.log("Unexpected response:", checkResponse);
            process.exit(1);
        }

        await new Promise(resolve => setTimeout(resolve, 10000)); // Wait for 10 seconds before checking again
    }
})();

File: nodenithy/run/public_key_service.py
import requests
import time
import argparse

BASE_URL = "https://publickey.ethernity.cloud"


def submit_ipfs_hash(
    hhash,
    enclave_name,
    protocol_version,
    network,
    template_version,
    docker_composer_hash,
):
    url = f"{BASE_URL}/api/addHash"
    payload = {
        "hash": hhash,
        "enclave_name": enclave_name,
        "protocol_version": protocol_version,
        "network": network,
        "template_version": template_version,
        "docker_composer_hash": docker_composer_hash,
    }
    response = requests.post(url, json=payload)
    print(f"response:{response}")
    if response.status_code == 200:
        return response.json()
    else:
        print("Error:", response.con)
        exit(1)


def check_ipfs_hash_status(hash):
    url = f"{BASE_URL}/api/checkHash/{hash}"
    response = requests.get(url)
    if response.status_code == 200:
        return response.json()
    else:
        print("Error:", response.json())
        exit(1)


if __name__ == "__main__":
    # Example usage
    # python3 ./public_key_service.py --enclave_name ${ENCLAVE_NAME_SECURELOCK} --protocol_version ${VERSION} --network ${BLOCKCHAIN_NETWORK} --template_version ${VERSION}

    # read enclave_name, protocol_version, network, template_version from command line arguments
    parser = argparse.ArgumentParser()
    parser.add_argument("--enclave_name", help="Enclave name", required=True)
    parser.add_argument("--protocol_version", help="Protocol version", required=True)
    parser.add_argument("--network", help="Network", required=True)
    parser.add_argument("--template_version", help="Template version", required=True)

    args = parser.parse_args()
    enclave_name = args.enclave_name
    protocol_version = args.protocol_version
    network = args.network.lower().split("_")[1]
    template_version = args.template_version

    hhash = ""
    with open("IPFS_HASH.ipfs", "r") as f:
        hhash = f.read().strip()
    docker_composer_hash = ""
    with open("IPFS_DOCKER_COMPOSE_HASH.ipfs", "r") as f:
        docker_composer_hash = f.read().strip()

    print("IPFS Hash:", hhash)
    print("Enclave Name:", enclave_name)
    print("Protocol Version:", protocol_version)
    print("Network:", network)
    print("Template Version:", template_version)
    print("Docker Composer Hash:", docker_composer_hash)

    # Submit IPFS Hash
    submit_response = submit_ipfs_hash(
        hhash,
        enclave_name,
        protocol_version,
        network,
        template_version,
        docker_composer_hash,
    )
    print("Submit IPFS Hash Response:", submit_response)

    # Check IPFS Hash Status
    while True:
        check_response = check_ipfs_hash_status(hhash)
        if "publicKey" in check_response:
            if check_response["publicKey"] == 0:
                print(
                    f"Public key not available yet. Queue position: {check_response.get('queuePosition', 'Unknown')}"
                )
            elif check_response["publicKey"] == -1:
                print("Hash is not derived from Eternity Cloud SDK.")
                exit(1)
            else:
                print("Public Key:", check_response["publicKey"])
                # save public key to file
                with open("PUBLIC_KEY.txt", "w") as f:
                    f.write(check_response["publicKey"])
                break
        else:
            print("Unexpected response:", check_response)
            exit(1)

        time.sleep(10)  # Wait for 10 seconds before checking again
